homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,274 @@
+"""EnaSolar solar inverter interface."""
+
+from __future__ import annotations
+
+from datetime import date, datetime
+import logging
+from typing import Callable
+
+import pyenasolar
+
+from homeassistant.components.sensor import (
+    STATE_CLASS_MEASUREMENT,
+    STATE_CLASS_TOTAL_INCREASING,
+    SensorEntity,
+)
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_NAME,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    ENERGY_KILO_WATT_HOUR,
+    EVENT_HOMEASSISTANT_START,
+    POWER_KILO_WATT,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+from homeassistant.core import CALLBACK_TYPE, callback
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.event import async_call_later
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CONF_CAPABILITY,
+    CONF_DC_STRINGS,
+    CONF_MAX_OUTPUT,
+    CONF_SUN_DOWN,
+    CONF_SUN_UP,
+    DEFAULT_SUN_DOWN,
+    DEFAULT_SUN_UP,
+    DOMAIN,
+    ENASOLAR_UNIT_MAPPINGS,
+    SCAN_DATA_MIN_INTERVAL,
+    SCAN_MAX_INTERVAL,
+    SCAN_METERS_MIN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add enasolar entry."""
+
+    # Use all sensors by default, but split them to have two update frequencies
+    hass_meter_sensors = []
+    hass_data_sensors = []
+    listeners = []
+
+    host = config_entry.data[CONF_HOST]
+
+    _LOGGER.info("Instantiate an EnaSolar Inverter at '%s'", host)
+    enasolar = pyenasolar.EnaSolar()
+
+    try:
+        await enasolar.interogate_inverter(host)
+    except Exception as conerr:
+        _LOGGER.error("Connection to EnaSolar Inverter '%s' failed (%s)", host, conerr)
+        raise PlatformNotReady from conerr
```

Comment:
Set this up in `async_setup_entry` in `__init__.py` and store it in `hass.data`.  There you can raise `ConfigEntryNotReady` and the integration will retry setup later if the device is offline

Reply:
Not sure I have understood this yet, will need to spend sometime thinking it through.  May dfall into the 'next version' category as well

Reply:
Deferring at this time.

Reply:
We don't want new integrations setup from config entries using PlatformNotReady since the user won't get any feedback in the UI when something goes wrong. When ConfigEntryNotReady is raised, the string is propagated to the UI layer and the user can see why the integration is failing to setup.

Reply:
I think I get what you mean but can you point me at an example of an integration that does this as I am still not confident I understand it.

Reply:
Check emonitor

Note that the ConfigEntryNotReady exception is automatically raised by the call to await coordinator.async_config_entry_first_refresh() if it fails on the first refresh.

Reply:
Thanks, I really appreciate your spending time helping me with this, I REALLY should not have started by using saj as a kind of template, it was not up to what is now expected in terms of a good implementation

Reply:
Have substituted exception

Reply:
OK I have added to  __init__.py's async_setup_entry but am unsure now what exception to raise in sensor.py.